### DISCUSSION TITLE
Filter on nested object field and nested array of objects field.

### DISCUSSION THREAD

**Discussion by gmourier** [meilisearch/product#225] [Feedback & Feature Proposal]
As we do not support filtering on nested objects fields, it forces users to denormalize the data.

e.g.

```json
{
    "ISBN": "979-8455737053",
    "title": "Twenty Thousand Leagues Under the Seas: (Classics Illustrated and Annotated)",
    "price": 12.65,
    "author": {
        "name": "Jules Verne",
        "shortBio": "Jules Verne, born on February 8, 1828 in Nantes and died on March 24, 1905 in Amiens, is a French writer, whose work is devoted to novels of adventure and anticipation."
    }
}
```

needs to be transformed to

```json
{
    "ISBN": "979-8455737053",
    "title": "Twenty Thousand Leagues Under the Seas: (Classics Illustrated and Annotated)",
    "price": 12.65,
    "author_name": "Jules Verne",
    "author_shortBio": "Jules Verne, born on February 8, 1828 in Nantes and died on March 24, 1905 in Amiens, is a French writer, whose work is devoted to novels of adventure and anticipation."
}
```

It would be interesting to express a filter on a nested field.

```json
{
    "filter": "author.name = 'Jules Verne'"
}
```

We could go further and also allow to filter on a field of objects contained in an array.

e.g. Having books containing multiple co-authors.

```json
[
        {
            "ISBN": "979-8455737053",
            "title": "Twenty Thousand Leagues Under the Seas: (Classics Illustrated and Annotated)",
            "price": 12.65,
            "co-authors": [
                {
                    "name": "Jules Verne",
                    "shortBio": "Jules Verne, born on February 8, 1828 in Nantes and died on March 24, 1905 in Amiens, is a French writer, whose work is devoted to novels of adventure and anticipation."
                },
                {
                    "name": "Secret Jules Verne's sidekick",
                    "shortBio": null
                }
            ]
        }
]
```

e.g. A filter may be expressed this way

```json
{
    "filter": "co-authors.name = 'Jules Verne'"
}
```

**Comment by fatihyildizhan** [comment]
This feature will be amazing one 👀

**Comment by jkdoshi** [comment]
YES! This would be great. I just added a similar comment here  - https://github.com/meilisearch/meilisearch/issues/2211#issuecomment-1085256264

**Reply by gmourier** [reply]
Hey @jkdoshi 👋

We haven't included support for array index notation for this [iteration](https://github.com/meilisearch/product/discussions/225#discussioncomment-2775234), but it's potentially something we could add in the future, depending on the needs we gather.

I invite you to create a dedicated discussion giving us more information on this need. Thank you 🙏

**Comment by gmourier** [comment]
Hey 👋

Guess what? It's now a reality! ✨

You can now use the dot-notation `.` to specify the use of a nested field in the settings definition or at search time.

e.g. `"filter": "company.name = 'Meilisearch'"` 

Our goal was to remove the hassle you encountered by being forced to flatten the nested fields at the root of your document on your end to get the full power of Meilisearch.

Don't hesitate to let us know what can be improved by [creating a new discussion](https://github.com/meilisearch/product/discussions/new).

Thank you for being part of the Meilisearch community.
